qiime2/core/validate.py

```diff
@@ -0,0 +1,30 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2021, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+class SemanticValidation:
```

Comment:
What if this object was per-partition?

Then the sorting of the entire system could be amortized, and we don't create this objects just to discard them moments later.